File: SeaBattle/SeaBattle.CSharp/DraggableShip.cs
namespace SeatBattle.CSharp
{
    public class DraggableShip : Ship
    {
        private DraggableShip(int length)
            : base(length)
        {
        }

        public Ship Source { get; private set; }

        public static DraggableShip From(Ship ship)
        {
            var draggableShip = new DraggableShip(ship.Length)
                                    {
                                        X = ship.X,
                                        Y = ship.Y,
                                        Orientation = ship.Orientation,
                                        Source = ship
                                    };

            return draggableShip;
        }

        public bool IsOrientationModified { get; set; }
    }
}

File: SeaBattle/SeaBattle.Cpp/DraggableShip.cpp
#include "stdafx.h"
#include "Ship.h"
#include "DraggableShip.h"

DraggableShip::DraggableShip(int length) : Ship(length)
{
};

DraggableShip^ DraggableShip::From(Ship ^ship)
{
	DraggableShip^ draggableShip = gcnew DraggableShip(ship->Length);
    draggableShip->X = ship->X;
    draggableShip->Y = ship->Y;
    draggableShip->Orientation = ship->Orientation;
    draggableShip->Source = ship;

	return draggableShip;
};